glean-core/android/src/main/java/mozilla/telemetry/glean/upload/Upload.kt

```diff
@@ -0,0 +1,105 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.telemetry.glean.upload
+
+import org.json.JSONObject
+import com.sun.jna.Structure
+import com.sun.jna.Pointer
+import com.sun.jna.Union
+import mozilla.telemetry.glean.Glean
+import mozilla.telemetry.glean.net.HeadersList
+import mozilla.telemetry.glean.rust.getRustString
+
+@Suppress("ClassNaming")
+@Structure.FieldOrder("tag", "uuid", "path", "body", "headers")
+internal class Upload_Body(
+    // NOTE: We need to provide defaults here, so that JNA can create this object.
+    @JvmField val tag: Byte? = null,
+    @JvmField val uuid: Pointer? = null,
+    @JvmField val path: Pointer? = null,
+    @JvmField val body: Pointer? = null,
+    @JvmField val headers: Pointer? = null
+) : Structure() {
+    fun toPingRequest(): PingRequest {
+        return PingRequest(
+            this.path!!.getRustString(),
+            this.body!!.getRustString(),
+            this.headers!!.getRustString()
+        )
+    }
+}
+
+@Structure.FieldOrder("tag", "upload")
+internal open class FfiPingUploadTask(
+    // NOTE: We need to provide defaults here, so that JNA can create this object.
+    @JvmField var tag: Byte = 2,
+    @JvmField var upload: Upload_Body = Upload_Body()
+) : Union() {
+    class ByValue : FfiPingUploadTask(), Structure.ByValue
+
+    fun toPingUploadTask(): PingUploadTask {
+        return when (this.tag.toInt()) {
+            0 -> PingUploadTask.Wait
+            1 -> {
+                this.readField("upload")
+                PingUploadTask.Upload(this.upload.toPingRequest())
+            }
+            else -> PingUploadTask.Done
+        }
+    }
+}
+
+/**
+ * The PingUploadTask makes it easier to consume Upload_Body.
+ */
+internal class PingRequest(
+    val path: String,
+    val body: String,
+    headers: String
+) {
+    val headers: HeadersList = headersFromJSONString(headers)
+
+    companion object {
+        fun headersFromJSONString(str: String): HeadersList {
+            val jsonHeaders = JSONObject(str)
```

Comment:
This doesn't seem to handle the [possible exception that could be raised here](https://developer.android.com/reference/kotlin/org/json/JSONObject#init_3): `org.json.JSONException` can get thrown if the string doesn't represent valid JSON for some reason.  Can you add a test to see what happens if the `str` passed in is invalid JSON?  At the very least we should probably log an error.

Reply:
As a side note: We generate the JSON this consumes inside Rust. The only way it can be invalid is if we get memory corruption when copying out the string.

In the principle of defensive coding it might still make sense to catch it, but comment that it's unlike to happen.

Reply:
If this has a low likelihood of occurring, then a comment works for me.  I've just been bitten by the not so terribly descriptive `JSONException` before.

Reply:
Should we even send a ping if we can't parse de headers? Isn't this the case where we have something like PingUploadTask::Skip or something of the sort? If that is the case I we could open a follow up for it. For now I will just log and return an empty headers list.

Reply:
I don't think we should attempt to send if we fail to parse. We should log an error.